atlasdb-dagger/src/main/java/com/palantir/atlasdb/services/RawKeyValueServiceModule.java

```diff
@@ -17,18 +17,56 @@
 
 import com.palantir.atlasdb.keyvalue.api.KeyValueService;
 import com.palantir.atlasdb.util.MetricsManager;
+import com.palantir.logsafe.logger.SafeLogger;
+import com.palantir.logsafe.logger.SafeLoggerFactory;
 import dagger.Module;
 import dagger.Provides;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.TimeUnit;
 import javax.inject.Named;
 import javax.inject.Singleton;
 
 @Module
 public class RawKeyValueServiceModule {
+    private static final SafeLogger log = SafeLoggerFactory.get(RawKeyValueServiceModule.class);
+    private static final ScheduledExecutorService executor = Executors.newSingleThreadScheduledExecutor();
+    private static final int DELAY_BETWEEN_INITIALIZATION_CHECKS_MILLIS = 100;
 
     @Provides
     @Singleton
     @Named("rawKvs")
     public KeyValueService provideRawKeyValueService(ServicesConfig config, MetricsManager metricsManager) {
         return config.atlasDbSupplier(metricsManager).getKeyValueService();
     }
+
+    @Provides
+    @Singleton
+    @Named("initializedRawKvs")
+    public KeyValueService provideInitializedRawKeyValueService(@Named("rawKvs") KeyValueService rawKeyValueService) {
+        return transformToFuture(rawKeyValueService).join();
+    }
+
+    private CompletableFuture<KeyValueService> transformToFuture(KeyValueService keyValueService) {
+        CompletableFuture<KeyValueService> future = new CompletableFuture<>();
+        ScheduledFuture<?> scheduledFuture = executor.scheduleAtFixedRate(
+                () -> {
+                    try {
+                        if (keyValueService.isInitialized()) {
+                            future.complete(keyValueService);
+                        } else {
+                            log.info("Waiting for KeyValueService to initialize");
+                        }
+                    } catch (Exception e) {
+                        future.completeExceptionally(e);
+                    }
+                },
+                0,
+                DELAY_BETWEEN_INITIALIZATION_CHECKS_MILLIS,
+                TimeUnit.MILLISECONDS);
+        future.whenComplete((_result, _thrown) -> scheduledFuture.cancel(true));
+        return future;
```

Comment:
How would we ensure that the executor is shut down once the computation is complete? It is just one executor, but we should try to keep things clean.